### PR TITLE
Auto generate index.routing_path from mapping

### PR DESCRIPTION
**Pull request by martijnvg**
If no `index.routing_path` has been specified in the matching template or the component template that the template refers to then attempt to generate the `index.routing_path` index setting from the matching template's mapping or component template the template refers to.

Relates #74660

### CODE REVIEW

File: modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamIndexSettingsProvider.java [hunk 1]
```diff
@@ -92,4 +118,54 @@ public Settings getAdditionalIndexSettings(
         return Settings.EMPTY;
     }
 
+    // Find fields in mapping that are of type keyword and time_series_dimension enabled.
```
Comment (nik9000) [inline modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamIndexSettingsProvider.java:121]: I'd use javadoc for this. It'll show up when you mouseover the method in your IDE which is kind of nice.

**Review by nik9000 — COMMENTED** [review]
It makes sense to me. I'm not particularly worried about the overhead involved when validating templates. I *think* it's a fairly infrequent action - just on template PUT. Is that true?

**Comment by martijnvg** [comment]
>  I think it's a fairly infrequent action - just on template PUT. Is that true?

This will be used when creating a template (as part of validating), when simulating a template via simulate api and creation of a backing index of data stream. The latter will be the most frequent compared to the other two use cases and happens as part of creating a new data stream or rolling over a data stream. I still think that this is okay.

**Comment by martijnvg** [comment]
Pinging @jsoriano for awareness.

**Review by nik9000 — APPROVED** [review]

**Pull request merged**